pkgs/stdenv/generic/setup.sh

```diff
@@ -562,6 +562,10 @@ _addToEnv() {
             (( "$depHostOffset" <= "$depTargetOffset" )) || continue
             local hookRef="${hookVar}[$depTargetOffset - $depHostOffset]"
             if [[ -z "${strictDeps-}" ]]; then
+
+                # Keep track of which packages we have visited before.
+                local visitedPkgs=""
```

Comment:
(I'm not an expert, but would it be better to use an array instead of building/scanning a string? IIRC there are some issues with some shell features so maybe it's a bad idea for those or other reasons)